app/views/unattended/provisioning_templates/registration/global_registration.erb

```diff
@@ -112,31 +117,70 @@ if [ x$ID = xrhel ] || [ x$ID = xcentos ]; then
 <%= "          --data 'remote_execution_interface=#{@remote_execution_interface}' \\\n" if @remote_execution_interface.present? -%>
 <%= "          --data 'packages=#{@packages}' \\\n" if @packages.present? -%>
 
-}
+    }
 
-    <% if @force -%>
-    yum remove -y katello-ca-consumer*
-    <% end -%>
+    KATELLO_SERVER_CA_CERT=/etc/rhsm/ca/katello-server-ca.pem
+    RHSM_CFG=/etc/rhsm/rhsm.conf
+
+    # Prepare SSL certificate
+    cp -f $SSL_CA_CERT $KATELLO_SERVER_CA_CERT
+    chmod 644 $KATELLO_SERVER_CA_CERT
 
-    CONSUMER_RPM=$(mktemp --suffix .rpm)
-    curl --silent --show-error --output $CONSUMER_RPM <%= subscription_manager_configuration_url(hostname: @url_host) %>
+    # Prepare subscription-manager
+    yum remove -y katello-ca-consumer*
 
-    # Workaround for systems with enabled FIPS,
-    # where installation of RPM generated on RHEL7 cause 'no digest' error
-    # See https://projects.theforeman.org/issues/32068
-    if [ "$(cat /proc/sys/crypto/fips_enabled)" = "1" ]; then
-        rpm -ivh --nodigest --nofiledigest $CONSUMER_RPM
+    if ! [ -x "$(command -v subscription-manager)" ] ; then
+      if [ "${VERSION_ID:0:1}" -gt 7 ]; then
+        dnf install -y subscription-manager
+      else
+        yum install -y subscription-manager
+      fi
     else
-        yum localinstall $CONSUMER_RPM -y
+      if [ "${VERSION_ID:0:1}" -gt 7 ]; then
+        dnf upgrade -y subscription-manager
+      else
+        yum upgrade -y subscription-manager
+      fi
     fi
 
-    rm -f $CONSUMER_RPM
+    if ! [ -f $RHSM_CFG ] ; then
+      echo "'$RHSM_CFG' not found, cannot configure subscription-manager"
+      cleanup_and_exit 1
+    fi
 
-    subscription-manager register <%= '--force' if @force %> --org='<%= @organization.label %>' --activationkey='<%= activation_keys %>' || <%= @ignore_subman_errors ? 'true' : 'exit 1' %>
+    # Configure subscription-manager
+    test -f $RHSM_CFG.bak || cp $RHSM_CFG $RHSM_CFG.bak
+    subscription-manager config \
+      --server.hostname="<%= @url.host %>" \
+      --server.port="<%= @url.port %>" \
```

Comment:
What does this port value resolve to? The reason I ask is that today, if a host is registering to Foreman this would be 443, if a host is registering to a content proxy this port would be 8443. We have a story to change this on a content proxy to be 443 and thus align it with the server. That change would make this registration story cleaner over all but we would need to land that change first. We are starting work on it very soon.

Reply:
Btw, this is the reason why I really am insisting that we use helpers for this. Maybe I should have made it even more explicit.

Reply:
Hmm, TBH totally forgot about smart-proxy! So what should be the behavior here? Do we want to `subscription-manager` call the proxy or ignore the proxy's URL & set `server.hostname` & `server.port` to Foreman?

Reply:
The proxy! Right now port 8443 for RHSM and port 443 for pulp, to be consolidated later into 443.

Reply:
I decided to spend a bit of time to come up with a partial patch. https://github.com/Katello/katello/pull/9413 has the following code:
```ruby
def rhsm_url
  # TODO: We shouldn't relate this to the Pulp feature and also expose it
  # as a setting
  if pulp_mirror?
    "https://#{URI.parse(url).host}:8443/rhsm"
  elsif pulp_primary?
    "https://#{URI.parse(url).host}/rhsm"
  end
end
```

So you can convert this to a host macro:
```ruby
def host_rhsm_url(host)
  return unless host.respond_to?(:content_source) && host.content_source

  smart_proxy = host.content_source
  # TODO: We shouldn't relate this to the Pulp feature and also expose it
  # as a setting
  if smart_proxy.pulp_mirror?
    "https://#{URI.parse(smart_proxy.url).host}:8443/rhsm"
  elsif smart_proxy.pulp_primary?
    "https://#{URI.parse(smart_proxy.url).host}/rhsm"
  end
end
```

Reply:
Hmm, not sure how to do it in the `global_registration` template.
* I cannot use `host_rhsm_url(host)` because I don't have `host` object yet (host is not created yet).
* The only way to get URL is from `url` parameter or take Foreman's default url [0]
* And I can't use `pulp_mirror? || pulp_primary?` because I don't know if the URL passed to `/register?url=...` [0] is smart-proxy's  URL or something else.

[0] https://github.com/theforeman/foreman/blob/develop/app/controllers/concerns/foreman/controller/registration.rb#L85-L96

Reply:
You do have a Smart Proxy instance right? That's your content source instance. If not, you should use SmartProxy::pulp3_primary I believe.

Note that to be really safe, you should verify the Pulp 3 feature is present on that Smart Proxy.